overrides: added howto docs (#470)
inveniosoftware/docs-invenio-rdm: docs/develop/howtos/override_components.md

> # How to override a react component

Reviewer: ```suggestion
# How to override a React component
```

> This documentation is targeted to developers that want to customize a specific react component in an instance.

Reviewer: ```suggestion
This documentation is targeted to developers that want to customize a specific React component in an instance.
```
and all other occurrences (search in ALL the doc?). Thanks!

## Overridable

> It is possible to override certain react components in Invenio RDM, by using the library [`react-overridable`](https://github.com/indico/react-overridable). This feature is available to instance developers to customize the look and feel of the react application (e.g. deposit form).

Reviewer: ```suggestion
It is possible to override certain react components in InvenioRDM, by using the library [`react-overridable`](https://github.com/indico/react-overridable). This feature is available to instance developers to customize the look and feel of the react application (e.g. deposit form).
```
and all other occurrences (search in ALL the doc?). Thanks!

> It is possible to override certain react components in Invenio RDM, by using the library [`react-overridable`](https://github.com/indico/react-overridable). This feature is available to instance developers to customize the look and feel of the react application (e.g. deposit form).

Reviewer: I would add a very short explanation of what you can do with `react-overridable` e.g. override React components by unique cmp ids.

> Despite the library allowing multiple ways of overriding a component, for this guide we will implement a new component to replace the default one.

Reviewer: I would suggest explaining here that you will override the `Metadata-only record` checkbox, to make clear what the next steps are about.

## Steps

- Make sure that the component you want to override is implemented in the same level as an `override.js` file. For example, for the deposit form, `invenio-app-rdm` has the following structure:

```terminal
├── invenio_app_rdm
│   ├── assets
│   │   ├── semantic-ui
│   │   |   ├── js
│   │   |   |   ├── invenio_app_rdm
│   │   |   |   |   ├── deposit
│   │   |   |   |   |   ├── override.js
│   │   |   |   |   |   ├── RDMDepositForm.js
```

We will override the file `override.js`.

- Find a react component to override. For this example, we are going to override the deposit form '`Metadata-only record`' checkbox.

!['`Metadata-only record`' checkbox](./img/metadata_only_checkbox.png)

!!! info "You must know the component id beforehand"
    In order to override a component, the exact overridable id must be retrieved beforehand.
    For now, you must retrieve the component id directly from the module's code.

> The react component's overridable id for the '`Metadata-only record`' checkbox is `ReactInvenioDeposit.MetadataAccess.layout`.

Reviewer 2: Info block is not rendering properly in the displayed documentation.
![Screenshot from 2023-01-16 11-11-05](https://user-images.githubusercontent.com/25476209/212653155-edbf7643-9800-4a3b-a1de-3bad63d41b26.png)

For this example, we will replace the '`Metadata-only record`' checkbox with another component, like a [Toggle](https://react.semantic-ui.com/modules/checkbox/#types-toggle) component.

> - Create a file named `override.js` with the same path as the one defined by the module (e.g. `<instance_name>/assets/js/invenio_app_rdm/deposit/override.js`). Note that `semantic-ui` was ommited.

Reviewer: ```suggestion
- Create a file named `override.js` with the same path as the one defined by the module (e.g. `<instance_name>/assets/js/invenio_app_rdm/deposit/override.js`). Note that `semantic-ui` was ommitted.
```

> - Create a file (e.g. `MetadataToggle.js`) to implement your component, under `<instance_name>/assets/js/components`. If the folder `components` does not exist, you can create it.

Reviewer 3: Suggestion: The new components do not have a requirement on where they have to be, I would maybe phrase it as a recommendation to keep/create a components folder, you could even add another level after it, to group the components like `.../components/deposit`.

Reviewer: What do you mean with "the linking problem"? We override by `id`, not by replacing the file.

```

- Create a react component to override the default one. In the file `MetadataToggle.js`, add the following code:

    ```javascript
    import React from "react";
    import { Checkbox } from "semantic-ui-react";
    import { useFormikContext } from "formik";
    import PropTypes from "prop-types";

    const MetadataToggle = (props) => {
    const { filesEnabled } = props;
    const { setFieldValue } = useFormikContext();

    const handleOnChangeMetadataOnly = () => {
        setFieldValue("files.enabled", !filesEnabled);
        setFieldValue("access.files", "public");
    };

    return (
        <Checkbox
        toggle
        label="Metadata-only record"
        onChange={handleOnChangeMetadataOnly}
        />
    );
    };

    export default MetadataToggle;

    MetadataToggle.propTypes = {
    filesEnabled: PropTypes.bool.isRequired,
    };
    ```

- Add the following object to `override.js`, specifying both components: the one to be replaced and its replacement.

```javascript
import MetadataToggle from "../../components/MetadataToggle";

export const overriddenComponents = {
  "ReactInvenioDeposit.MetadataOnlyToggle.layout": MetadataToggle,
};
```

> - Lastly, rebuild your assets. In a terminal, run the following command:

Reviewer 4: ```suggestion
- Lastly, rebuild your assets - the new file needs to be added to your instance. In a terminal, run the following command:
```
maybe these small justification would help people to learn on which command should be run when

  
```terminal

> invenio-cli assets build

Reviewer: Very very minor: I would add before:
```
cd my-site
```

```

- The final result shows the override component: 

!["`Metadata-only record`" toggle](./img/metadata_only_toggle.png)

## Other examples

> 

Reviewer 3: Question: I am wondering if we could add a link to a public repository who is implementing this, it might be usefull to "see it in action".
For intance [cds-rdm](https://github.com/CERNDocumentServer/cds-rdm) is using it .
We could link it and mention that there are some real examples available there, if it makes sense to you.

Author: I thought about providing the examples of Zenodo and CDS-RDM (I was following them myself when writing the docs). But I share the concern with @kpsherva  that it might change at some point, leaving the docs with a broken link.

Reviewer: It looks like that the link of @jrcastro2 is pointing to the GitHub project homepage, it should be ok no? Agree that a link to a file will break...
Maybe better to add a ZenodoRDM one?

Reviewer 5: I will add a link to zenodo-rdm repository in case people want to see more examples in action.
